diff --git a/meson.build b/meson.build
index ec6f9b4..24829d8 100644
--- a/meson.build
+++ b/meson.build
@@ -32,7 +32,7 @@ cfg = configuration_data()
 cfg.set_quoted('PACKAGE_VERSION', meson.project_version())
 
 include_dirs = [ include_directories('.') ]
-sshfs_sources = ['sshfs.c', 'cache.c']
+sshfs_sources = ['sshfs.c', 'cache.c', 'cache2.c']
 if target_machine.system() == 'darwin'
   cfg.set_quoted('IDMAP_DEFAULT', 'user')
   sshfs_sources += [ 'compat/fuse_opt.c', 'compat/darwin_compat.c' ]
diff --git a/sshfs.c b/sshfs.c
index 5513266..04d9488 100644
--- a/sshfs.c
+++ b/sshfs.c
@@ -52,6 +52,7 @@
 #endif
 
 #include "cache.h"
+#include "cache2.h"
 
 #ifndef MAP_LOCKED
 #  define MAP_LOCKED 0
@@ -471,6 +472,10 @@ enum {
 #define SSHFS_OPT(t, p, v) { t, offsetof(struct sshfs, p), v }
 
 static struct fuse_opt sshfs_opts[] = {
+	/* for enabling fuse_fs::debug */
+	FUSE_OPT_KEY("debug", FUSE_OPT_KEY_KEEP),
+	FUSE_OPT_KEY("-d", FUSE_OPT_KEY_KEEP),
+
 	SSHFS_OPT("directport=%s",     directport, 0),
 	SSHFS_OPT("ssh_command=%s",    ssh_command, 0),
 	SSHFS_OPT("sftp_server=%s",    sftp_server, 0),
@@ -4338,6 +4343,10 @@ int main(int argc, char *argv[])
 	if (res == -1)
 		exit(1);
 
+	res = cache2_parse_options(&args);
+	if (res == -1)
+		exit(1);
+
 	sshfs.randseed = time(0);
 
 	if (sshfs.max_read > 65536)
@@ -4355,6 +4364,7 @@ int main(int argc, char *argv[])
 		sshfs.op = cache_wrap(&sshfs_oper);
 	else
 		sshfs.op = &sshfs_oper;
+	sshfs.op = cache2_wrap(sshfs.op);
 	fuse = fuse_new(&args, sshfs.op,
 			sizeof(struct fuse_operations), NULL);
 	if(fuse == NULL)
